### ISSUE TITLE
`EmitterWriter::get_max_line_num` works incorrectly

### ISSUE THREAD

**Issue by AnthonyMikh** [rust-lang/rust#64447]
**TL;DR:** [`EmitterWriter::get_max_line_num`](https://github.com/rust-lang/rust/blob/master/src/librustc_errors/emitter.rs#L1037) unconditionally returns `self.get_multispan_max_line_num(span)` no matter what is contained in `children`.

Longer explanation: let's take a look at the code of method (it's rather short):
```rust
    fn get_max_line_num(&mut self, span: &MultiSpan, children: &[SubDiagnostic]) -> usize {
        let mut max = 0;

        let primary = self.get_multispan_max_line_num(span);
        max = if primary > max { primary } else { max };

        for sub in children {
            let sub_result = self.get_multispan_max_line_num(&sub.span);
            max = if sub_result > max { primary } else { max };
        }
        max
    }
```
Here `self.get_multispan_max_line_num(span)` returns a plain `usize`. Firstly, since `0` is the smallest possible value of `usize`, the first three lines can be rewritten without changing the meaning as
```rust
let primary = self.get_multispan_max_line_num(span);
let mut max = primary;
```
So after executing these lines `max == primary`. Secondly, in the loop `if` compares `sub_result` with `max` but assigns either `max` or `primary`. If `max == primary` in the beginning of iteration, then this also holds in the end of iteration. Since this preposition holds before executing the loop, it also holds after executing the loop, so the whole method just returns `primary`, QED.

Erroneous code was introduced in 71ec2867e3c4cc448cb64890daffd1d5ffbe353b. 

Unfortunately, I'm not familliar enough with the code to propose the correct fix.

**Comment by Centril** [comment]
cc @estebank @phansch

**Comment by AnthonyMikh** [comment]
Closing as fixed by #64721
